### ISSUE TITLE
featuresAt callback delay after adding geoJSONSource at high zoom level

### ISSUE THREAD

**Issue by kristfal** [mapbox/mapbox-gl-js#1507]
Hey,

`featuresAt` callback may at times be as slow as 70+ seconds after adding a geoJSONSource at high zoom levels. The source data I'm testing with is a line string of 12 kB, and it renders instantly. 

This issue only happens if the geoJSON is loaded at high zoom levels. At low zoom levels, there is no delay. During this delay, CPU activity is 100%.

Adding source data with (can't add through URL in this use case):

```javascript
sourceObj = new mapboxgl.GeoJSONSource({
  data: ls.data
});
this.mgl.addSource(ls.source, sourceObj);
```

Tap event:

```javascript

_mapTap: function (e) {
      var x = e.detail.x, 
          y = e.detail.y,
          self = this,
          featureId = null,
          latlng = this.mgl.unproject([x, y]);
  
      this.mgl.featuresAt([x,y], { radius: 15 }, function (err, features) {
        
        if (features[0]) {
          featureId = features[0].properties.id;  
        } else {
          featureId = null;
        }
        
        return self.tapevent = {
          x: x,
          y: y,
          featureId: featureId,
          latlng: latlng
        };
      });
}

```

After the initial wait/delay, `featuresAt` triggers instantly again. If multiple tap events are triggered during the delay, they queue up and fire in sequence. `unproject` and other methods trigger instantly at any time during this delay.

I can only reproduce this issue if the source is loaded a high zoom level (12+). Loading the source at lower zoom levels and triggering a click event doesn't cause a delay, nor does zooming in and clicking after a load.

Profiling this issue leads me to believe geoJSONVT.splitTile is causing the slowdown. It seems to me like the issue only happens if the geoJSON crosses tile boundaries upon initialization, but I can't verify this 100%.

See attached screenshot for profiling results. Let me know if you need any other info.

![skjermbilde 2015-09-21 14 37 46](https://cloud.githubusercontent.com/assets/5638107/9991985/162f1d14-606f-11e5-849b-1ddbf005b0d1.png)

**Comment by jfirebaugh** [comment]
@kristfal Can you post an executable test case on JSFiddle? Also, can you post the profiler `.cpuprofile` result file?

**Comment by kristfal** [comment]
@jfirebaugh Working duplication at https://jsfiddle.net/paoqrwec/5/  

See inline comments for workarounds, let me know if it doesn't lock up for you.

**Comment by mourner** [comment]
Wow, that looks wild. So apparently it's only reproducible with an empty GeoJSON, right? Investigating.

**Comment by kristfal** [comment]
@mourner That is correct.

It also requires a high zoom level when the GeoJSON is loaded. If you change it to 2 in the fiddle and re-run, it will work even with the empty GeoJSON.

**Comment by mourner** [comment]
Confirmed to be an upstream geojson-vt bug, https://github.com/mapbox/geojson-vt/issues/40

**Comment by mourner** [comment]
Released geojson-vt 2.1.6 with the fix. Should be included in the build of the next gl-js version. Also discovered a different bug that led to more tiles being generated than necessary, fixed as well. Thanks for the report!

**Comment by kristfal** [comment]
@mourner Thanks for the fast turnaround, I'll keep you posted if I find anything else after it has trickled downstream.
